7/10 cleaned up voting

samiklein011011 opened the pull request:
Cleaned up a lot of the files and made more generalizable functions. The one file I haven't finished cleaning up yet is the /ajax/ajaxUndoVote.php function, but I'm not sure if maybe there is a better way to do rather than take in this ifUndoAndVote parameter. 

Thanks!

Commit:
6/19 first page of events website!

Good first draft. most of the content is on the page, displayed fairly nicely. don't have columns working over the picture yet, need to do that.

Commit:
6/19, did work on trending events page!

got 2 columns, and 2 columns within each box. woo!!!

Commit:
6/20 stopping point

got a lot done!!!
- connected the trending events and single events pages
- trending events is organized by the most votes first
- little rectangles on trending events page with info
- got a 1 2 3 and then lots of 3 rows on trending events page. and used % so that it only prints the exact number of cards needed based upon events
- the single event page has eveything needed from the database
- changed around a lot of the CSS. it looks pretty decent (although it's lime and yellow right now, mostly for my own purposes, and it looks kinda crazy)

Commit:
6/21 before master commit

commiting here before merging to master

Commit:
6/21, index.php file in root now

Commit:
6/21 fixed index and trending page i think

let's see what happens

Commit:
6/21

getting there.

created log in stuff. basically just copy & pasted from blog tho. also attempted to do javascript on the uparrows today. tried to use a jquery plug in (which is what those files are). But didn't work too well.

Commit:
6/21 upvote works!!!!!!!

That is all

Commit:
6/25 upVote&downVote

upVote&downVote are both working on a single event page, but not working on trending events yet.

Commit:
6/25 organization

not important rlly, just got rid of some unessecary variables

Commit:
6/25 tried a bunch of stuff

trying to make the JD function recallable, and get thte voting to work on the trending page, but failing

Commit:
6/29

i have no idea what i'm doing w voting. good.

Commit:
6/29 i have no idea what's going on, sadly

sad friday commit vibes

Commit:
7/2 voting generally works, the corrrect way

voting generally works, the corrrect way.

each event has a different id on the votes, so can call it specifically.

the 1 problem is that it jumps to the top of the page on the trending page when it votes, but that's a problem for another day (aka probably after lunch lol)

Commit:
7/2 good stopping point

great day!!
finished voting (except one minor thing),
got categories to work in the form,
starting to make it so that the person who created the event can edit their event

Commit:
7/3 before sabotage

wooo!

Commit:
committed by Tyler

This is sabotage.

Commit:
7/3 fixed the sabotage!

mwahaha

Commit:
7/6 this is kinda working

kinda working.
but u can toggle between 4 numbers. that's not ok. i need to seperate upvoting and switching back to normal. so like that project we did in 247. with the 4 transformations. this can either be regular upvote, regular downvote, or back to normal+regUpVote or  back to normal+regDownVote.

but wanted to committ before i screw everyting up.

or but also - the gray isnt working on the trending events page - it changes the correct event number, but the incorrect event color (always changes event #1 color

Commit:
7/6 i think voting works on the UI single event page!

ditto

Commit:
7/9 voting works completely on the single events page, but not the trending events page

^^ title

good monday morning.

Commit:
7/9 i think voting works everywhere!

wow. on both client and server side, and on both single event page and trending events page. woooooo i can move on yayyy

Commit:
7/9 just some organization, small stuff

- Put all of the files in the eventsProject folder just straight into htdocs, so could make trendingEvents page = index.php (also did that).

-fixed password visitbility

- made username unique

- trying to build out friends thing (now u can see each others profiles)

Commit:
7/10 clean-up of functions

- created a seperate file for JS functions
- cleaned up most of the functions

Commit:
7/10 style, minor fixes

- users who aren't logged in can't vote (on both the trending events page and the single events page frame)
- some style fixes

tilikang commented on include/echoFunctions.php:
```diff
@@ -0,0 +1,104 @@
+		</head>
+
+";
+include('include/jsFunctions.php');
```
This works, but what I was talking about earlier was that you could make a `.js` file and include it with a script tag.

tilikang commented on include/jsFunctions.php:
```diff
@@ -0,0 +1,124 @@
+<!-- <?php
+
+$userID=$_SESSION['userID'];
```
just so you're aware, since you're including this file in the header (meaning in lots of different places) and this is setting `$userID` in the global scope, that means you're creating this variable all over the place in your code, not just for this file. This might not be what you intended to do.

If you want this to be contained just to this file, wrap the whole thing in a function. If you'd rather have $userID available everywhere (probably not a bad idea) I think it should go somewhere more general, not in your jsFunctions.php file.

tilikang commented on include/jsFunctions.php:
```diff
@@ -0,0 +1,124 @@
+		jsVote(eventID, direction);
+	}
+
+	else if((getCurrentVoteDirection(eventID)=="down" && direction=="down") ||(getCurrentVoteDirection(eventID)=="up" && direction == "up")){
```
You could simplify this logic by just saying

```
else if(getCurrentVoteDirection(eventID) == direction)){
```

There may be opportunities to use this simplification elsewhere in your code. Not just in if statements, but as a general approach to how you're thinking about all the different cases you need to handle.

tilikang commented on include/jsFunctions.php:
```diff
@@ -0,0 +1,124 @@
+	console.log("before you voted = " +total);
+
+	if(direction == "up"){
+		document.getElementById("upVoteButton_"+eventID).classList.add('voted');
```
This is another place you could use the trick I mentioned above. There's a way to do this without having a separate case for up and down voting. Just say `document.getElementById(direction+"VoteButton_"+eventID).classList.add('voted');`

tilikang commented on include/jsFunctions.php:
```diff
@@ -0,0 +1,124 @@
+		console.log("after you voted = " +(total+1));
+	}
+
+	$.post({ url:'/ajax/ajaxUndoVote.php', data:{eventID, direction, ifUndoAndVote}});
```
I would only do an ajax call to undo this if there isn't an additional vote being registered. Right now, if you're doing an undo and vote at the same time, it sends two ajax calls and it's possible they'll reach the server in the wrong order which would give incorrect results.

The server can realize that it needs to undo a past vote when the new one comes in. You don't need to explicitly tell it to do the undo.

tilikang commented on ajax/ajaxUndoVote.php:
```diff
@@ -0,0 +1,42 @@
+$sessionUserID=$_SESSION['userID'];
+$userVote = doesUserVoteExist($eventID, $sessionUserID);
+
+deleteUserVote($userVote['0']['eventID'], $_SESSION['userID']);
```
You might as well check that the userVote exists before calling this.

You can put this exact code into the endpoint that processes a new vote, so you won't need to call this undo endpoint at all in that case (it'll automatically undo any existing votes before registering the new one)

tilikang commented on ajax/ajaxUndoVote.php:
```diff
@@ -0,0 +1,42 @@
+	}
+}
+else{	//if this is just undo-ing the vote, no next steps
+	if($_REQUEST['direction'] == "up"){
```
This is minor, but I'd probably make a `voteInDB` function that takes a direction. That way calling it will be cleaner, and you won't need this if statement here (it'll go in the function instead, and it's normally better to hide complexity in functions so this code is cleaner)

tilikang replied:
Also, since this is the undo endpoint, I don't think you need to handle voting here if you take my other advice,

tilikang commented on ajax/ajaxUndoVote.php:
```diff
@@ -0,0 +1,42 @@
+
+include('config/init.php');
+
+if(!isset($_SESSION['userID'])) {  //&& !empty($_SESSION['userID']) <-- had this before. not sure if necessary
```
You could make this a common function. Like `checIfUserIsLoggedIn`

tilikang commented on ajax/ajaxVote.php:
```diff
@@ -0,0 +1,33 @@
+//also assumed 3 cases: 1) insert&vote 2)delete&vote and 3)delete, insert, and vote
+if($userVote == null){
+	insertUserVote($eventID, $sessionUserID, $direction);
+	generalVoteinDB($eventID, $direction);
```
Why'd you use the `downVoteInDB` function in the undo endpoint, but here you have a general function? I think these two endpoints might need to be reconciled.

tilikang commented on ajax/ajaxVote.php:
```diff
@@ -0,0 +1,33 @@
+
+
+else if(($userVote['0']['direction'] == "down" && $direction == "down")||($userVote['0']['direction'] == "up" && $direction == "down")){
+	deleteUserVote($userVote['0']['eventID'], $sessionUserID);
```
This works, but I'd just handle undos separate from this whole if clause. At the top of this file, just see if the user has already voted on this event, and if they have deleted it. Then, just insert the new vote (without regard for whether an undo just happened). Separating them out will simplify this.

tilikang reviewed (commented):
@samikleinLACRM - This looks a lot cleaner. I think there's still some room for improvement, but it's getting pretty lean which is great.

Commit:
7/11 after PR

- fixing up the PR from yesterday
- I'm not really sure why this is cleaner, it looks like just as much code as yesterday, and just as complicated

Commit:
7/12 voting with commented out code

i think this is pretty solid?

This commit has all of the commented out code that I'm too scared to delete

Commit:
7/12 voting cleaned up

- got rid of commented out stuff

- i think voting is pretty clean?

Commit:
7/12 end of the day

- if someone isn't logged in, a lil message pops up saying that their vote cannot be saved
- editProfile now works correctly

Commit:
7/13 morning

- trying to work on being able to edit the alert button
